include/envoy/registry/extensions_registry.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+#include "envoy/registry/registry.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/substitute.h"
+
+namespace Envoy {
+namespace Registry {
+
+/**
+ * Helper class to call `registeredNames` for a given type.
+ */
+class BaseExtensionFactoryRegistry {
+public:
+  virtual ~BaseExtensionFactoryRegistry() = default;
+  virtual std::vector<absl::string_view> registeredNames() const PURE;
+};
+
+template <class Base> class ExtensionFactoryRegistry : public BaseExtensionFactoryRegistry {
+public:
+  using FactoryRegistry = Envoy::Registry::FactoryRegistry<Base>;
+
+  std::vector<absl::string_view> registeredNames() const {
+    return FactoryRegistry::registeredNames();
+  }
+};
+
+/**
+ * This is a registry of extension factory registries. The distinction between
+ * a factory registry and an extension factory registry is that an extension
+ * factory registry registers factories that are visible to user configuration
+ * through the build system and API.
+ */
+class ExtensionFactoryRegistryRegistry {
+public:
+  using MapType = absl::flat_hash_map<absl::string_view, BaseExtensionFactoryRegistry*>;
```

Comment:
As implemented, using string_view as key is safe. It would make sleep better though if this was std::string.

Reply:
Fixed.